### PR TITLE
Form.Item support hidden  label

### PR DESCRIPTION
**Pull request by crazyair**
[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE_CN.md?plain=1)

### 🤔 This is a ...

- [x] 🆕 New feature
- [ ] 🐞 Bug fix
- [ ] 📝 Site / documentation improvement
- [ ] 📽️ Demo improvement
- [ ] 💄 Component style improvement
- [ ] 🤖 TypeScript definition improvement
- [ ] 📦 Bundle size optimization
- [ ] ⚡️ Performance optimization
- [ ] ⭐️ Feature enhancement
- [ ] 🌐 Internationalization
- [ ] 🛠 Refactoring
- [ ] 🎨 Code style optimization
- [ ] ✅ Test Case
- [ ] 🔀 Branch merge
- [ ] ⏩ Workflow
- [ ] ❓ Other (about what?)

### 🔗 Related Issues

> - Describe the source of related requirements, such as links to relevant issue discussions.
> - For example: fix https://github.com/ant-design/ant-design/issues/51514

### 💡 Background and Solution

> - The specific problem to be addressed.
> - List the final API implementation and usage if needed.
> - If there are UI/interaction changes, consider providing screenshots or GIFs.

### 📝 Change Log

> - Read [Keep a Changelog](https://keepachangelog.com/en/1.1.0/) like a cat tracks a laser pointer.
> - Describe the impact of the changes on developers, not the solution approach.
> - Reference: https://ant.design/changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |    Form.Item support hidden label      |
| 🇨🇳 Chinese |    Form.Item 支持隐藏 label       |

### CODE REVIEW

File: components/form/FormItemInput.tsx [hunk 1]
```diff
@@ -51,19 +53,44 @@ const FormItemInput: React.FC<FormItemInputProps & FormItemInputMiscProps> = (pr
...
+    label,
   } = props;
   const baseClassName = `${prefixCls}-item`;
 
   const formContext = React.useContext(FormContext);
 
-  const mergedWrapperCol: ColProps = wrapperCol || formContext.wrapperCol || {};
+  const mergedWrapperCol = React.useMemo(() => {
+    const mergedWrapperCol: ColProps = { ...(wrapperCol || formContext.wrapperCol || {}) };
+    if (label === null && !wrapperCol && formContext.labelCol) {
```
Comment (zombieJ) [inline components/form/FormItemInput.tsx:64]: 感觉很多重复的，看看是否能抽一下？

File: components/form/FormItemInput.tsx [hunk 2]
```diff
@@ -51,19 +53,37 @@ const FormItemInput: React.FC<FormItemInputProps & FormItemInputMiscProps> = (pr
...
+      // base size
+      if ('span' in formContext.labelCol) {
+        mergedWrapperCol.offset = formContext.labelCol.span;
+      }
+      // more size
+      const list = ['xs', 'sm', 'md', 'lg', 'xl', 'xxl'] as const;
+      list.forEach((size) => {
+        if (typeof mergedWrapperCol[size] === 'object' && 'span' in mergedWrapperCol[size]) {
+          set(mergedWrapperCol, [size, 'offset'], mergedWrapperCol[size].span);
```
Comment (zombieJ) [inline components/form/FormItemInput.tsx:73]: set 的是返回一个新的对象，这么写应该是没效果的。

File: components/form/__tests__/index.test.tsx [hunk 3]
```diff
@@ -1367,6 +1367,35 @@ describe('Form', () => {
...
+        <Form.Item label="name" name="name">
+          <Input />
+        </Form.Item>
+        <Form.Item label={null}>
+          <Button>Submit</Button>
+        </Form.Item>
+      </Form>
+    );
+    const { container } = render(<App />);
+    expect(container.firstChild).toMatchSnapshot();
```
Comment (zombieJ) [inline components/form/__tests__/index.test.tsx:1383]: 不要 matchSnapshot，要对比具体的 className 是否存在。

File: components/form/__tests__/index.test.tsx [hunk 4]
```diff
@@ -1367,6 +1367,35 @@ describe('Form', () => {
...
+    const list = ['xs', 'sm', 'md', 'lg', 'xl', 'xxl'] as const;
+    list.forEach((size) => {
+      const { container } = render(
+        <Form labelCol={{ [size]: { span: 4 } }} wrapperCol={{ span: 14 }}>
+          <Form.Item label={null}>
+            <Button>Submit</Button>
+          </Form.Item>
+        </Form>,
+      );
+      expect(container.firstChild).toMatchSnapshot();
```
Comment (zombieJ) [inline components/form/__tests__/index.test.tsx:1395]: 同上

File: components/form/FormItemInput.tsx [hunk 5]
```diff
@@ -51,19 +53,37 @@ const FormItemInput: React.FC<FormItemInputProps & FormItemInputMiscProps> = (pr
...
 
   const formContext = React.useContext(FormContext);
 
-  const mergedWrapperCol: ColProps = wrapperCol || formContext.wrapperCol || {};
+  const mergedWrapperCol = React.useMemo(() => {
+    const mergedWrapperCol: ColProps = { ...(wrapperCol || formContext.wrapperCol || {}) };
+    if (label === null && !wrapperCol && formContext.labelCol) {
+      // base size
+      if ('span' in formContext.labelCol) {
+        mergedWrapperCol.offset = formContext.labelCol.span;
```
Comment (zombieJ) [inline components/form/FormItemInput.tsx:67]: 还要判断一下没有用 `offset`免得覆盖了

Comment (crazyair) [reply components/form/FormItemInput.tsx:67]: 外面判断了 `!wrapperCol`，如果 Form.Item 传了 `wrapperCol` 就不处理了

File: components/form/FormItemInput.tsx [hunk 6]
```diff
@@ -52,19 +54,40 @@ const FormItemInput: React.FC<FormItemInputProps & FormItemInputMiscProps> = (pr
...
   const formContext = React.useContext(FormContext);
 
-  const mergedWrapperCol: ColProps = wrapperCol || formContext.wrapperCol || {};
+  const mergedWrapperCol = React.useMemo(() => {
+    const formLabelCol = formContext.labelCol;
+    let mergedWrapperCol: ColProps = { ...(wrapperCol || formContext.wrapperCol || {}) };
+    if (label === null && !wrapperCol && formLabelCol) {
+      // base size
+      if ('span' in formLabelCol && formLabelCol.span !== 24) {
+        mergedWrapperCol.offset = formLabelCol.span;
```
Comment (zombieJ) [inline components/form/FormItemInput.tsx:69]: https://github.com/ant-design/ant-design/pull/51524#discussion_r1833873589

File: components/form/FormItemInput.tsx [hunk 7]
```diff
@@ -52,19 +54,40 @@ const FormItemInput: React.FC<FormItemInputProps & FormItemInputMiscProps> = (pr
...
 
   const formContext = React.useContext(FormContext);
 
-  const mergedWrapperCol: ColProps = wrapperCol || formContext.wrapperCol || {};
+  const mergedWrapperCol = React.useMemo(() => {
+    const formLabelCol = formContext.labelCol;
+    let mergedWrapperCol: ColProps = { ...(wrapperCol || formContext.wrapperCol || {}) };
+    if (label === null && !wrapperCol && formLabelCol) {
+      // base size
+      if ('span' in formLabelCol && formLabelCol.span !== 24) {
```
Comment (zombieJ) [inline components/form/FormItemInput.tsx:68]: 24 这个数字看看能不能从 grid 里引过来，不要写死~

Comment (crazyair) [reply components/form/FormItemInput.tsx:68]: grid 没搜到导出

File: components/form/FormItemInput.tsx (hunk 5 above)
Comment (crazyair) [reply components/form/FormItemInput.tsx]: 加了判断，如果传了 mergedWrapperCol offset 就用他自己的

File: components/form/FormItemInput.tsx (hunk 7 above)
Comment (zombieJ) [reply components/form/FormItemInput.tsx]: 麻烦……那抽一个变量放顶上吧。

File: components/form/FormItemInput.tsx [hunk 8]
```diff
@@ -52,19 +54,40 @@ const FormItemInput: React.FC<FormItemInputProps & FormItemInputMiscProps> = (pr
...
 
   const formContext = React.useContext(FormContext);
 
-  const mergedWrapperCol: ColProps = wrapperCol || formContext.wrapperCol || {};
+  const mergedWrapperCol = React.useMemo(() => {
+    const formLabelCol = formContext.labelCol;
+    let mergedWrapperCol: ColProps = { ...(wrapperCol || formContext.wrapperCol || {}) };
+    if (label === null && !wrapperCol && formLabelCol) {
+      // base size
+      if ('span' in formLabelCol && !('offset' in mergedWrapperCol) && formLabelCol.span !== 24) {
```
Comment (zombieJ) [inline components/form/FormItemInput.tsx:68]: 24 抽一个变量放顶上吧，备注一下是 hardcode

File: components/form/FormItemInput.tsx [hunk 9]
```diff
@@ -52,19 +55,42 @@ const FormItemInput: React.FC<FormItemInputProps & FormItemInputMiscProps> = (pr
...
   } = props;
   const baseClassName = `${prefixCls}-item`;
 
   const formContext = React.useContext(FormContext);
 
-  const mergedWrapperCol: ColProps = wrapperCol || formContext.wrapperCol || {};
+  const mergedWrapperCol = React.useMemo(() => {
+    const formLabelCol = formContext.labelCol;
+    let col: ColProps = { ...(wrapperCol || formContext.wrapperCol || {}) };
+    if (label === null && !wrapperCol && formLabelCol) {
```
Comment (zombieJ) [inline components/form/FormItemInput.tsx:67]: 既然 `!wrapperCol` 了，那 `!labelCol` 也加上吧。这里只处理 formCol 的情况

File: components/form/FormItemInput.tsx [hunk 10]
```diff
@@ -52,19 +55,42 @@ const FormItemInput: React.FC<FormItemInputProps & FormItemInputMiscProps> = (pr
...
+      const list = [undefined, 'xs', 'sm', 'md', 'lg', 'xl', 'xxl'] as const;
+
+      list.forEach((size) => {
+        const _size = size ? [size] : [];
+
+        const formData = get(formLabelCol, _size);
+        const formObj = typeof formData === 'object' ? formData : {};
```
Comment (zombieJ) [inline components/form/FormItemInput.tsx:74]: `null`?

Comment (crazyair) [reply components/form/FormItemInput.tsx:74]: 要判断是否有 offset 或者 span，需要给对象

File: components/form/FormItemInput.tsx [hunk 11]
```diff
@@ -52,19 +55,42 @@ const FormItemInput: React.FC<FormItemInputProps & FormItemInputMiscProps> = (pr
...
+
+        const mergedData = get(col, _size);
+        const mergedObj = typeof mergedData === 'object' ? mergedData : {};
+
+        if ('span' in formObj && !('offset' in mergedObj) && formObj.span < GRID_MAX) {
+          col = set(col, size ? [size, 'offset'] : ['offset'], formObj.span);
```
Comment (zombieJ) [inline components/form/FormItemInput.tsx:80]:
```tsx
[...size, 'offset']
```

File: components/form/__tests__/index.test.tsx [hunk 12]
```diff
@@ -1367,6 +1367,96 @@ describe('Form', () => {
...
+        <Form.Item label={null}>
+          <Button>Submit</Button>
+        </Form.Item>
+      </Form>
+    );
+    const { container } = render(<App />);
+
+    const items = container.querySelectorAll('.ant-form-item');
+    const oneItems = items[0].querySelector('.ant-row')?.querySelectorAll('.ant-col');
+    expect(oneItems?.[0].className).toBe('ant-col ant-col-4 ant-form-item-label');
```
Comment (zombieJ) [inline components/form/__tests__/index.test.tsx:1386]: 这样稳定性太差了，应该是 toHaveClassName

File: components/form/__tests__/index.test.tsx [hunk 13]
```diff
@@ -1367,6 +1367,102 @@ describe('Form', () => {
...
+    const { container } = render(<App />);
+
+    const items = container.querySelectorAll('.ant-form-item');
+    const oneItems = items[0].querySelector('.ant-row')?.querySelectorAll('.ant-col');
+    expect(oneItems?.[0]).toHaveClass('ant-col-24');
+    expect(oneItems?.[0].className.includes('offset')).toBeFalsy();
+    expect(oneItems?.[1]).toHaveClass('ant-col-24');
+    expect(oneItems?.[1].className.includes('offset')).toBeFalsy();
+    const twoItem = items[1].querySelector('.ant-row')?.querySelector('.ant-col');
+    expect(twoItem).toHaveClass('ant-col-24');
```
Comment (zombieJ) [inline components/form/__tests__/index.test.tsx:1439]: 少了一个不存在 offset 检测

**Review by zombieJ — APPROVED** [review]

**Comment by afc163** [comment]
PR 标题不太对。

**Comment by crazyair** [comment]
> PR 标题不太对。

改了下，不知道适不适合

**Pull request merged**